File: backend/shared/preflight.ts
import { logger } from "./logger";

export type ServiceCheck = {
  /** Human-readable service name, shown in the startup error. */
  name: string;
  /** Lightweight call that rejects if the service is unavailable. */
  check: () => Promise<unknown>;
};

/**
 * Runs every health check and aborts startup if any third-party service is
 * unavailable. Checks run concurrently so a single boot reports *all* failing
 * services at once rather than one at a time.
 */
export async function verifyServices(checks: ServiceCheck[]): Promise<void> {
  const results = await Promise.allSettled(checks.map((c) => c.check()));

  const failures = results.flatMap((result, i) =>
    result.status === "rejected"
      ? [{ name: checks[i].name, reason: result.reason }]
      : [],
  );

  if (failures.length > 0) {
    for (const { name, reason } of failures) 
      logger.error(`Service unavailable at startup: ${name}`, reason);
    
    const detail = failures
      .map(({ name, reason }) => {
        const message = reason instanceof Error ? reason.message : String(reason);
        return `  - ${name}: ${message}`;
      })
      .join("\n");
    throw new Error(
      `Startup aborted — the following services are unavailable:\n${detail}`,
    );
  }
}

File: youtube-fetcher/app/preflight.py
"""Fail-fast preflight, mirroring backend/shared/preflight.ts: run every
health check and abort startup if any dependency is unavailable. Unlike the
backend's concurrent Promise.allSettled, checks run sequentially — but every
failure is still collected so one boot reports all unavailable services."""

import logging
from typing import Callable

log = logging.getLogger(__name__)

# (human-readable name, lightweight call that raises if unavailable)
ServiceCheck = tuple[str, Callable[[], object]]


def verify_services(checks: list[ServiceCheck]) -> None:
    failures: list[tuple[str, Exception]] = []

    for name, check in checks:
        try:
            check()
        except Exception as error:
            log.error("Service unavailable at startup: %s: %s", name, error)
            failures.append((name, error))

    if failures:
        detail = "\n".join(f"  - {name}: {error}" for name, error in failures)
        raise RuntimeError(
            f"Startup aborted — the following services are unavailable:\n{detail}"
        )
